core/lib/dal/src/consensus_dal.rs

```diff
@@ -30,54 +30,129 @@ impl ConsensusDal<'_, '_> {
+    pub async fn try_update_genesis(&mut self, genesis: &validator::Genesis) -> anyhow::Result<()> {
+        let mut txn = self.storage.start_transaction().await?;
+        if let Some(got) = txn.consensus_dal().genesis().await? {
+            // Exit if the genesis didn't change.
+            if &got == genesis {
+                return Ok(());
+            }
+            anyhow::ensure!(
+                got.fork.number < genesis.fork.number,
+                "transition to a past fork is not allowed: old = {:?}, new = {:?}",
+                got.fork.number,
+                genesis.fork.number,
+            );
+            anyhow::ensure!(
+                got.fork.first_parent.is_none(),
+                "fork with first_parent != None not supported",
+            );
+        }
+        let genesis =
+            zksync_protobuf::serde::serialize(genesis, serde_json::value::Serializer).unwrap();
+        let state = zksync_protobuf::serde::serialize(
+            &ReplicaState::default(),
+            serde_json::value::Serializer,
+        )
+        .unwrap();
+        sqlx::query!(
+            r#"
+            DELETE FROM miniblocks_consensus
+            "#
+        )
+        .execute(txn.conn())
+        .await?;
+        sqlx::query!(
+            r#"
+            DELETE FROM consensus_replica_state
+            "#
+        )
+        .execute(txn.conn())
+        .await?;
         sqlx::query!(
             r#"
             INSERT INTO
-                consensus_replica_state (fake_key, state)
+                consensus_replica_state (fake_key, genesis, state)
             VALUES
-                (TRUE, $1)
-            ON CONFLICT (fake_key) DO
-            UPDATE
-            SET
-                state = excluded.state
+                (TRUE, $1, $2)
             "#,
-            state
+            genesis,
+            state,
         )
-        .execute(self.storage.conn())
+        .execute(txn.conn())
         .await?;
+        txn.commit().await?;
         Ok(())
     }
 
-    /// Fetches the first consensus certificate.
-    /// Note that we didn't backfill the certificates for the past miniblocks
-    /// when enabling consensus certificate generation, so it might NOT be the certificate
-    /// for the genesis miniblock.
-    pub async fn first_certificate(&mut self) -> anyhow::Result<Option<validator::CommitQC>> {
-        let Some(row) = sqlx::query!(
+    /// [Main node only] creates a new consensus fork starting at
+    /// the last sealed miniblock. Resets the state of the consensus
+    /// by calling `try_update_genesis()`.
+    pub async fn fork(&mut self) -> anyhow::Result<()> {
+        let mut txn = self.storage.start_transaction().await?;
+        let Some(old) = txn.consensus_dal().genesis().await? else {
+            return Ok(());
+        };
+        let last = txn.blocks_dal().get_sealed_miniblock_number().await?;
+        let first_block = validator::BlockNumber(last.map(|n| n.0.into()).unwrap_or(0));
```

Comment:
- Why is having no miniblocks in the storage OK here? Why is the default value zero? What about snapshot recovery? (At least in the decentralized network, and maybe even before that, we'll enable snapshot recovery for main node as well.)
- Repeating the responsibility question from the line comment above: Why should it be responsibility of DAL (vs the calling app) to decide whether having no miniblocks in the storage is OK? If this would be the app's responsibility, we could reason that e.g. consensus components never run with empty storage (not saying that it necessarily makes sense, just that app-level reasoning wouldn't have to be lowered to the DAL level).

Reply:
* here you go, disabled forking with no blocks
* as long as multiple independent components access the same storage through dal, it is responsibility of dal to enforce invariants. Otherwise one component has no guarantee that the other component didn't mess up the state. It is a really bad idea to scatter the responsibility for maintaining state invariants over a large codebase. Precedence that it is already the case doesn't decrease the value of the effort of moving invariants to the place where they can be actually enforced.